### PR TITLE
Multi module enhancements

### PR DESCRIPTION
**Pull request by dshimo**
## Description
This PR has grown quite significantly, so this description seeks to breakdown as many changes. This description is handwritten by me. An overall summary is provided at the bottom.

## constants.ts
`contextValues` have been updated to include 'tags' for `isContainer`, `isAggregator`, and for Maven and Gradle.
The specific Maven/Gradle commands for dev mode are keyed out. The registered commands, defined in the `package.json`, were added.

## libertyProject.ts
This file grew quickly by essentially becoming the one-stop shop for building Liberty Projects and rendering it. To help organize these responsibilities, this file strictly dictates how the tree item is rendered and its various commands, such as terminal creation. The large diff is created from a mass subtraction.

Project crawling/discovery, the tree render, the tree storage, are all moved to `projectDiscovery.ts`, `projectTreeProvider`, and `projectRegistry` respectively.

## projectDiscovery
This file has the most significant changes. The core of its idea are as follows:
- Initiate a parallel crawl for workspace folders. By asynchronously starting this, we hope to update the view responsively, and correct tree associations as we read more.
- "Discover" phase: Identify all build files, and read them for 'modules' definitions and such. 
- "Stamp" phase: Passed the information from the discover phase to assign labels, such as isAggregator, to the respective Liberty objects
- "Link" phase: Strictly creates a map to link parents with their children

`addServerXMLProjects` acts as the fallback to find lonely server.xml files

## projectTreeProvider
This implements VS Code's TreeDataProvider API. We have the standard `getChildren()`, and critically added `pickProject()`. Pick project implements the quick pick feature that will allow us to apply smart filters and determine which project to operate on, be it aggregator or direct module.

The part for `LibertyDevDecorationProvider` hides the Liberty status in the URI, to allow us to associate the item with a green visual.

## projectRegistry
This acts as the single source of truth for Liberty Projects for our tree view. 

## extension.ts
The LS waits were changed to `Promise.all`, to allow the wait to be in parallel instead of sequential. The command registrations were optionally squashed, and we additionally took the free `collapseAll` command to add to our view.

## devCommands
showProjects is deleted in favor of pickProject.
ensureTerminal was written as a consistent pattern for creating or getting the terminal for a given project.
COMMAND_TITLES map associates the defined commands to localized titles.
stopDevMode calls terminal dispose and clears isDevMode for tracking

## mavenUtil
Originally, `pluginManagement` was checked. This is incomplete behavior as pluginManagement dictates versions for its modules and doesn't actually enable the plugin.
Additionally, there was a typo to be corrected (`result.project.arifactId`)

## commandUtils
getCommandForMaven() is updated to append the artifactId to commands to support multi-module
getCommandForGradle() had a hardcoded path for the gradle wrapper. Updated to use an absolute path

## helperUtil
filterProjects() is rewritten. It supports the various isBoolean checks and additionally filters for state awareness, i.e. isDevMode. If a project is running, that project is not listed when running the start command. 

## package.json
Build file command is added, along with an icon. Menu entries are added to aggregators.

## Additional Unit Testing
To confirm edge cases, truth tables were made to better confirm different permutations of parent and child relations. There are Maven and Gradle tests to account for this.

---

## tl;dr
LibertyProject split into projectDiscovery, projectRegistry, projectTreeProvider.
This separates the significant logic for file discovery from the logic for rendering the UI elements.

The contextValue mechanism is reworked for regex string comparisons.

Files, LS, are made parallel when it makes sense to.

isDevMode tags if a module is running. We apply filtering logic and also show a visual indication of a running server in the view.

### CODE REVIEW

File: src/locales/en.json [hunk 1]
```diff
@@ -72,15 +72,20 @@
...
     "command.start.dev.mode": "start dev mode",
     "command.start.container.dev.mode": "start container dev mode",
     "command.start.custom.dev.mode": "start custom dev mode",
     "command.stop.dev.mode": "stop dev mode",
     "command.run.tests": "run tests",
     "command.attach.debugger": "attach debugger",
-    "command.view.test.report": "view test report"
+    "command.view.test.report": "view test report",
+    "scanning.workspace.folders": "Scanning workspace\u2026 ({0} of {1} folders)",
+    "scanning.workspace.complete": ""
```
Comment (mattbsox) [inline src/locales/en.json]: Needed if empty?

File: src/liberty/devCommands.ts [hunk 2]
```diff
@@ -23,133 +25,115 @@ class LibertyProjectQuickPickItem implements QuickPickItem {
...
+    }
+    if (terminal !== undefined) {
+        terminal.show();
+        targetProject.setTerminal(terminal);
+        if (isMaven(targetProject.getContextValue())) {
+            const pomPath = targetProject.parent ? targetProject.parent.getPath() : targetProject.getPath();
+            const artifactId = (targetProject.parent && (clickedDirectlyOnChild || ProjectRegistry.getInstance().findLibertyDescendants(targetProject.parent).length > 1))
+                ? targetProject.artifactId
+                : undefined;
+            const cmd = await getCommandForMaven(pomPath, "io.openliberty.tools:liberty-maven-plugin:dev", targetProject.getTerminalType(), undefined, artifactId);
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Pull the dev mode command into a constant

File: src/liberty/devCommands.ts [hunk 3]
```diff
@@ -23,133 +25,115 @@ class LibertyProjectQuickPickItem implements QuickPickItem {
...
+            terminal.sendText(cmd);
+        } else if (isGradle(targetProject.getContextValue())) {
+            const cmd = await getCommandForGradle(targetProject.getPath(), "libertyDev", targetProject.getTerminalType());
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Pull the Gradle command into a constant too

File: src/liberty/devCommands.ts [hunk 4]
```diff
@@ -353,273 +326,233 @@ export async function attachDebugger(libProject?: LibertyProject | undefined): P
...
+            ? Path.dirname(targetProject.parent.getPath())
+            : Path.dirname(targetProject.getPath());
+        terminal = createTerminalforLiberty(targetProject, terminal, terminalPath);
+    }
+    if (terminal !== undefined) {
+        terminal.show();
+        targetProject.setTerminal(terminal);
+
+        let placeHolderStr = "";
+        let promptString = localize("specify.custom.parms.maven");
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Typo

File: src/liberty/devCommands.ts [hunk 5]
```diff
@@ -353,273 +326,233 @@ export async function attachDebugger(libProject?: LibertyProject | undefined): P
...
+        if (isMaven(targetProject.getContextValue())) {
+            placeHolderStr = "e.g. -DhotTests=true";
+        } else if (isGradle(targetProject.getContextValue())) {
+            placeHolderStr = "e.g. --hotTests";
+            promptString = localize("specify.custom.parms.gradle");
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Same typo

File: src/liberty/devCommands.ts [hunk 6]
```diff
@@ -353,273 +326,233 @@ export async function attachDebugger(libProject?: LibertyProject | undefined): P
...
                 const pomPath = targetProject.parent ? targetProject.parent.getPath() : targetProject.getPath();
-                // Use module selector (-pl) if:
-                // 1. User clicked directly on child module, OR
-                // 2. Parent has multiple Liberty children
-                const artifactId = (targetProject.parent && (clickedDirectlyOnChild || projectProvider.findLibertyDescendants(targetProject.parent).length > 1))
+                const artifactId = (targetProject.parent && (clickedDirectlyOnChild || registry.findLibertyDescendants(targetProject.parent).length > 1))
                     ? targetProject.artifactId
                     : undefined;
-                const cmd = await getCommandForMaven(pomPath, "io.openliberty.tools:liberty-maven-plugin:devc", targetProject.getTerminalType(), undefined, artifactId);
+                const cmd = await getCommandForMaven(pomPath, "io.openliberty.tools:liberty-maven-plugin:dev", targetProject.getTerminalType(), customCommand, artifactId);
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Externalize this string.

File: src/liberty/devCommands.ts [hunk 7]
```diff
@@ -353,273 +326,233 @@ export async function attachDebugger(libProject?: LibertyProject | undefined): P
...
                 terminal.sendText(cmd);
-            } else if (targetProject.getContextValue() === LIBERTY_GRADLE_PROJECT_CONTAINER) {
-                const cmd = await getCommandForGradle(targetProject.getPath(), "libertyDevc", targetProject.getTerminalType());
+            } else if (isGradle(targetProject.getContextValue())) {
+                const cmd = await getCommandForGradle(targetProject.getPath(), "libertyDev", targetProject.getTerminalType(), customCommand);
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Same as above.

File: src/liberty/devCommands.ts [hunk 8]
```diff
@@ -353,273 +326,233 @@ export async function attachDebugger(libProject?: LibertyProject | undefined): P
...
-                : Path.dirname(targetProject.getPath());
-            //function call to create new terminal for LTV
-            terminal = createTerminalforLiberty(targetProject, terminal, terminalPath);
-        }
-        if (terminal !== undefined) {
-            terminal.show();
-            targetProject.setTerminal(terminal);
-            if (targetProject.getContextValue() === LIBERTY_MAVEN_PROJECT_CONTAINER) {
-                // For multi-module projects, use parent aggregator path
+            if (isMaven(targetProject.getContextValue())) {
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Can add a container project check here.

File: src/liberty/devCommands.ts [hunk 9]
```diff
@@ -353,273 +326,233 @@ export async function attachDebugger(libProject?: LibertyProject | undefined): P
...
+    }
+    if (terminal !== undefined) {
+        terminal.show();
+        targetProject.setTerminal(terminal);
+        if (isContainer(targetProject.getContextValue()) && isMaven(targetProject.getContextValue())) {
+            const pomPath = targetProject.parent ? targetProject.parent.getPath() : targetProject.getPath();
+            const artifactId = (targetProject.parent && (clickedDirectlyOnChild || ProjectRegistry.getInstance().findLibertyDescendants(targetProject.parent).length > 1))
+                ? targetProject.artifactId
+                : undefined;
+            const cmd = await getCommandForMaven(pomPath, "io.openliberty.tools:liberty-maven-plugin:devc", targetProject.getTerminalType(), undefined, artifactId);
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Externalize string.

File: src/liberty/devCommands.ts [hunk 10]
```diff
@@ -353,273 +326,233 @@ export async function attachDebugger(libProject?: LibertyProject | undefined): P
...
+            terminal.sendText(cmd);
+        } else if (isContainer(targetProject.getContextValue()) && isGradle(targetProject.getContextValue())) {
+            const cmd = await getCommandForGradle(targetProject.getPath(), "libertyDevc", targetProject.getTerminalType());
```
Comment (mattbsox) [inline src/liberty/devCommands.ts]: Same here.

File: src/liberty/projectDiscovery.ts [hunk 11]
```diff
@@ -0,0 +1,552 @@
+/*
+ * IBM Confidential
+ * Copyright IBM Corp. 2020, 2026
```
Comment (mattbsox) [inline src/liberty/projectDiscovery.ts:3]: Copyright

File: src/liberty/projectDiscovery.ts [hunk 12]
```diff
@@ -0,0 +1,552 @@
...
+			return {
+				path: p, type: "gradle",
+				parsedBuild,
+				settingsPath: settingsPath || undefined,
+				parsedSettings,
+				regexBuildFile: (!needsG2js && regexResult) ? regexResult : undefined,
+			};
+		}),
+	]);
+	console.log(`[perf] discoverProjects phase1 (read+parse): ${Date.now() - t0}ms`);
```
Comment (mattbsox) [inline src/liberty/projectDiscovery.ts:143]: Externalize this string if the message is needed.

Comment (mattbsox) [reply src/liberty/projectDiscovery.ts:143]: There are a lot of these in the rest of the file.

File: src/liberty/projectDiscovery.ts [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+	context: vscode.ExtensionContext,
+	projectsMap: Map<string, LibertyProject>,
+	existingProjects: Map<string, LibertyProject>
+): Promise<void> {
+	let serverXMLPaths: string[] = [];
+	const wsFolders = vscode.workspace.workspaceFolders;
+	if (wsFolders) {
+		for (const folder of wsFolders) {
+			if (!projectRootPathExists(folder.uri.fsPath, projectsMap.keys())) {
+				const pattern = new vscode.RelativePattern(folder.uri.fsPath, "**/src/main/liberty/config/server.xml");
```
Comment (mattbsox) [inline src/liberty/projectDiscovery.ts:439]: This could be configured in the build plugins as well.

File: src/liberty/projectDiscovery.ts [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+				const paths = (await vscode.workspace.findFiles(pattern, EXCLUDED_DIR_PATTERN)).map(u => u.fsPath);
+				serverXMLPaths = serverXMLPaths.concat(paths);
+			}
+		}
+	}
+
+	for (const serverXML of serverXMLPaths) {
+		const folder = vscodePath.parse(vscodePath.resolve(serverXML, "../../../../")).dir;
```
Comment (mattbsox) [inline src/liberty/projectDiscovery.ts:447]: If the configDir is set through the build plugins, this might not be the path to the project root.

File: src/liberty/projectRegistry.ts [hunk 15]
```diff
@@ -0,0 +1,165 @@
+/*
+ * IBM Confidential
+ * Copyright IBM Corp. 2020, 2026
```
Comment (mattbsox) [inline src/liberty/projectRegistry.ts:3]: Copyright

File: src/liberty/projectTreeProvider.ts [hunk 16]
```diff
@@ -0,0 +1,299 @@
+/*
+ * IBM Confidential
+ * Copyright IBM Corp. 2020, 2026
```
Comment (mattbsox) [inline src/liberty/projectTreeProvider.ts:3]: Copyright

File: src/liberty/projectTreeProvider.ts [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+				if (foldersComplete < totalFolders) {
+					this.setMessage(localize("scanning.workspace.folders", foldersComplete, totalFolders));
+				}
+				this._onDidChangeTreeData.fire(undefined);
+			}
+		);
+
+		this._registry.setProjects(projects);
+		this._registry.setRootProjects(rootProjects);
+		console.log(`[perf] updateProjects total: ${Date.now() - t0}ms  (${projects.size} projects, ${rootProjects.length} roots)`);
```
Comment (mattbsox) [inline src/liberty/projectTreeProvider.ts:302]: Can externalize this if the message is needed.

**Review by mattbsox — CHANGES_REQUESTED** [review]
Some minor things here, along with the other items we were thinking could be refactored that were discussed outside of this review.

File: src/liberty/projectDiscovery.ts (hunk 13 above)
Comment (dshimo) [reply src/liberty/projectDiscovery.ts:439]: So this appears in the PR since I moved the file from the refactor. But the logic was introduced back in 2022 with a commit SHA `3834c38`. I'll make an issue for enhancement.

File: src/liberty/projectDiscovery.ts (hunk 14 above)
Comment (dshimo) [reply src/liberty/projectDiscovery.ts:447]: Created issue #619 to cover this enhancement.

File: src/liberty/projectTreeProvider.ts [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+	 * This method asks the user to save the workspace first if it is untitled and contains
+	 * more than one project.
+	 */
+	public async checkUntitledWorkspaceAndSaveIt(): Promise<void> {
+		return new Promise((resolve) => {
+			try {
+				vscode.window.showInformationMessage(
+					localize("workspace.not.saved.projects.may.not.persist"),
+					{ modal: true },
+					'Save Workspace'
```
Comment (mattbsox) [inline src/liberty/projectTreeProvider.ts:134]: Should this be externalized? Maybe fine if this is the current implementation. Can update outside this PR if needed too.

File: src/liberty/projectDiscovery.ts (hunk 12 above)
Comment (mattbsox) [reply src/liberty/projectDiscovery.ts:143]: Can these `perf` and `discovery` scope messages be debug if they're needed?

**Review by mattbsox — CHANGES_REQUESTED** [review]
Seeing an issue detecting the aggregator and displaying the project as a multi-module project. The command needs to run `-pl <project-selector> -am -f <aggregator pom.xml>`.
```
 mvn io.openliberty.tools:liberty-maven-plugin:dev -pl :sample-ear -am -f "aggregator/pom.xml"
``` 

We can pull this into a new issue.

Related to #511

**Review by mattbsox — APPROVED** [review]
Looks good! Thanks!
